pkg/cmd/helmify.go

```diff
@@ -0,0 +1,144 @@
+
+//go:embed resources/secret.yaml
+var secretManifest []byte
+
+//go:embed resources/scaledobject.yaml
+var scaledObjectManifest []byte
+
+//go:embed resources/NOTES.txt
+var releaseNotes []byte
+
+//go:embed resources/_helpers.tpl
+var helpers []byte
+
+var helmifyCmd = &cobra.Command{
+	Use:   "helmify",
+	Short: "Scaffold a Helm chart for your SpinApp",
+	PreRunE: func(cmd *cobra.Command, args []string) error {
+		if len(scaffoldOpts.output) == 0 {
+			scaffoldOpts.output = defaultChartOutput
+		}
+		return nil
+	},
+	RunE: func(cmd *cobra.Command, args []string) error {
+		name, err := getNameFromImageReference(scaffoldOpts.from)
+		if err != nil {
+			return err
+		}
+		return scaffoldHelmChart(name, scaffoldOpts)
+	},
+}
+
+func scaffoldHelmChart(name string, opts ScaffoldOptions) error {
+	values, err := generateHelmValues(opts)
+	if err != nil {
+		return err
+	}
+	buf := new(bytes.Buffer)
+	err = values.Encode(buf)
+	if err != nil {
+		return fmt.Errorf("could not encode values for Helm chart")
+	}
+	c := chart.Chart{
+		Metadata: &chart.Metadata{
+			Name:        name,
+			Description: fmt.Sprintf("A Helm chart for %s", name),
+			Type:        "application",
+			APIVersion:  chart.APIVersionV2,
+			AppVersion:  "0.1.0",
+			Version:     "0.1.0",
+		},
+		Templates: []*chart.File{
+			{
+				Name: "values.yaml",
+				Data: buf.Bytes(),
+			},
+			{
+				Name: "templates/_helpers.tpl",
+				Data: []byte(helpers),
+			},
+			{
+				Name: "templates/app.yaml",
+				Data: []byte(appManifest),
+			},
+			{
+				Name: "templates/hpa.yaml",
+				Data: []byte(hpaManifest),
+			},
+			{
+				Name: "templates/scaledobject.yaml",
+				Data: []byte(scaledObjectManifest),
+			},
+			{
+				Name: "templates/secret.yaml",
+				Data: []byte(secretManifest),
+			},
+			{
+				Name: "templates/NOTES.txt",
+				Data: []byte(releaseNotes),
+			},
+			{
+				Name: ".helmignore",
+				Data: []byte(helmIgnore),
+			},
+		},
+		Values: values,
+	}
+	target := filepath.Base(opts.output)
+	return chartutil.SaveDir(&c, target)
+}
+
+func generateHelmValues(opts ScaffoldOptions) (chartutil.Values, error) {
+	values := chartutil.Values{
+		"image":            opts.from,
+		"replicas":         opts.replicas,
+		"cpuLimit":         opts.cpuLimit,
+		"memoryLimit":      opts.memoryLimit,
+		"cpuRequest":       opts.cpuRequest,
+		"memoryRequest":    opts.memoryRequest,
+		"imagePullSecrets": opts.imagePullSecrets,
+		"autoscaler":       opts.autoscaler,
+	}
+	if opts.configfile != "" {
+		raw, err := os.ReadFile(opts.configfile)
+		if err != nil {
+			return nil, err
+		}
+		values["runtimeConfig"] = base64.StdEncoding.EncodeToString(raw)
+	}
+	return values, nil
+}
+
+func init() {
+	// overwriting the default value of a flag defined on the parrent comand does not work
+	// that's why I went for a PreRunE
+	helmifyCmd.Flags().StringVarP(&scaffoldOpts.output, "out", "o", defaultChartOutput, "Path for writing the Helm chart to")
+	scaffoldCmd.AddCommand(helmifyCmd)
```

Comment:
I think it'd be better if we make this `spin kube helmify` rather than `spin kube scaffold helmify`. While both `scaffold` and `helmify` perform the same rough amount of work (prepare a Spin application to be deployed to Kubernetes), the use cases for both are quite distinct, and I think they should be treated as such.

By tying this command as a subcommand of `scaffold`, every feature flag that gets introduced to `scaffold` must now be added to `helmify`. There will be cases in the future where there will be a feature flag that will only work with `scaffold` but not `helmify`, and it'd be confusing for users to see that feature flag available in `helmify`'s help output because of the way this command has been architected.